### ISSUE TITLE
[BUG] BlazingContext is not releasing GPU memory

### ISSUE THREAD

**Issue by bsuryadevara** [BlazingDB/blazingsql#310]
**Describe the bug**
BlazingSQL context is not releasing acquired gpu memory though we call drop table. fyi, I'm running this on rapidsai docker container.
dataset location: http://files.grouplens.org/datasets/movielens/ml-20m.zip

**Steps/Code to reproduce bug**
```
import time
from blazingsql import BlazingContext

config = {"tables":[{"table_name":"movies", "filepath":"/rapids/ml/movies.csv"}]}

def run_query(query): 
    bc = BlazingContext()
    for table_config in config['tables']:
        table_name = table_config['table_name']
        file_path = table_config['filepath']
        bc.create_table(table_name, file_path)
    #Run query and get the results
    df = bc.sql(query)
    for table_config in config['tables']:
        bc.drop_table(table_config['table_name'])
        print("dropped table %s" %(table_config['table_name']))
    del bc

if __name__ == "__main__":
    for i in range(0, 10):
        run_query('SELECT * FROM main.movies LIMIT 100')
        print("iteration %s completed!"%(i))
        print("it's time to wait for 20secs")
        time.sleep(20)
```

**GPU:0 memory is constantly occupied**
```
+-----------------------------------------------------------------------------+
| Processes:                                                       GPU Memory |
|  GPU       PID   Type   Process name                             Usage      |
|=============================================================================|
+-----------------------------------------------------------------------------+
Wed Jan 29 09:21:56 2020       
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 418.87.00    Driver Version: 418.87.00    CUDA Version: 10.1     |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|===============================+======================+======================|
|   0  Tesla V100-SXM2...  On   | 00000000:07:00.0 Off |                    0 |
| N/A   34C    P0    43W / 300W |      0MiB / 32480MiB |      0%      Default |
+-------------------------------+----------------------+----------------------+
|   1  Tesla V100-SXM2...  On   | 00000000:85:00.0 Off |                    0 |
| N/A   33C    P0    56W / 300W |  28021MiB / 32480MiB |      0%      Default |
+-------------------------------+----------------------+----------------------+
|   2  Tesla V100-SXM2...  On   | 00000000:89:00.0 Off |                    0 |
| N/A   38C    P0    60W / 300W |  21171MiB / 32480MiB |      0%      Default |
+-------------------------------+----------------------+----------------------+
                                                                               
+-----------------------------------------------------------------------------+
| Processes:                                                       GPU Memory |
|  GPU       PID   Type   Process name                             Usage      |
|=============================================================================|
+-----------------------------------------------------------------------------+
Wed Jan 29 09:21:57 2020       
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 418.87.00    Driver Version: 418.87.00    CUDA Version: 10.1     |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|===============================+======================+======================|
|   0  Tesla V100-SXM2...  On   | 00000000:07:00.0 Off |                    0 |
| N/A   34C    P0    56W / 300W |   5308MiB / 32480MiB |      7%      Default |
+-------------------------------+----------------------+----------------------+
|   1  Tesla V100-SXM2...  On   | 00000000:85:00.0 Off |                    0 |
| N/A   33C    P0    56W / 300W |  28021MiB / 32480MiB |      0%      Default |
+-------------------------------+----------------------+----------------------+
|   2  Tesla V100-SXM2...  On   | 00000000:89:00.0 Off |                    0 |
| N/A   38C    P0    60W / 300W |  21171MiB / 32480MiB |      0%      Default |
+-------------------------------+----------------------+----------------------+
                                                                               
+-----------------------------------------------------------------------------+
| Processes:                                                       GPU Memory |
|  GPU       PID   Type   Process name                             Usage      |
|=============================================================================|
+-----------------------------------------------------------------------------+
Wed Jan 29 09:21:58 2020       
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 418.87.00    Driver Version: 418.87.00    CUDA Version: 10.1     |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|===============================+======================+======================|
|   0  Tesla V100-SXM2...  On   | 00000000:07:00.0 Off |                    0 |
| N/A   34C    P0    56W / 300W |  16828MiB / 32480MiB |      3%      Default |
+-------------------------------+----------------------+----------------------+
|   1  Tesla V100-SXM2...  On   | 00000000:85:00.0 Off |                    0 |
| N/A   33C    P0    56W / 300W |  28021MiB / 32480MiB |      0%      Default |
+-------------------------------+----------------------+----------------------+
|   2  Tesla V100-SXM2...  On   | 00000000:89:00.0 Off |                    0 |
| N/A   38C    P0    60W / 300W |  21171MiB / 32480MiB |      0%      Default |
+-------------------------------+----------------------+----------------------+
                                                                               
+-----------------------------------------------------------------------------+
| Processes:                                                       GPU Memory |
|  GPU       PID   Type   Process name                             Usage      |
|=============================================================================|
+-----------------------------------------------------------------------------+
Wed Jan 29 09:22:00 2020       
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 418.87.00    Driver Version: 418.87.00    CUDA Version: 10.1     |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|===============================+======================+======================|
|   0  Tesla V100-SXM2...  On   | 00000000:07:00.0 Off |                    0 |
| N/A   34C    P0    56W / 300W |  16828MiB / 32480MiB |      0%      Default |
+-------------------------------+----------------------+----------------------+
|   1  Tesla V100-SXM2...  On   | 00000000:85:00.0 Off |                    0 |
| N/A   33C    P0    56W / 300W |  28021MiB / 32480MiB |      0%      Default |
+-------------------------------+----------------------+----------------------+
|   2  Tesla V100-SXM2...  On   | 00000000:89:00.0 Off |                    0 |
| N/A   38C    P0    60W / 300W |  21171MiB / 32480MiB |      0%      Default |
+-------------------------------+----------------------+----------------------+
                                                                               
+-----------------------------------------------------------------------------+
| Processes:                                                       GPU Memory |
|  GPU       PID   Type   Process name                             Usage      |
|=============================================================================|
+-----------------------------------------------------------------------------+
Wed Jan 29 09:22:01 2020       
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 418.87.00    Driver Version: 418.87.00    CUDA Version: 10.1     |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|===============================+======================+======================|
|   0  Tesla V100-SXM2...  On   | 00000000:07:00.0 Off |                    0 |
| N/A   35C    P0    56W / 300W |  16828MiB / 32480MiB |      0%      Default |
+-------------------------------+----------------------+----------------------+
|   1  Tesla V100-SXM2...  On   | 00000000:85:00.0 Off |                    0 |
| N/A   33C    P0    56W / 300W |  28021MiB / 32480MiB |      0%      Default |
+-------------------------------+----------------------+----------------------+
|   2  Tesla V100-SXM2...  On   | 00000000:89:00.0 Off |                    0 |
| N/A   38C    P0    60W / 300W |  21171MiB / 32480MiB |      0%      Default |
+-------------------------------+----------------------+----------------------+
                                                                               
+-----------------------------------------------------------------------------+
| Processes:                                                       GPU Memory |
|  GPU       PID   Type   Process name                             Usage      |
|=============================================================================|
+-----------------------------------------------------------------------------+
Wed Jan 29 09:22:02 2020       
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 418.87.00    Driver Version: 418.87.00    CUDA Version: 10.1     |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|===============================+======================+======================|
|   0  Tesla V100-SXM2...  On   | 00000000:07:00.0 Off |                    0 |
| N/A   35C    P0    56W / 300W |  16828MiB / 32480MiB |      0%      Default |
+-------------------------------+----------------------+----------------------+
|   1  Tesla V100-SXM2...  On   | 00000000:85:00.0 Off |                    0 |
| N/A   33C    P0    56W / 300W |  28021MiB / 32480MiB |      0%      Default |
+-------------------------------+----------------------+----------------------+
|   2  Tesla V100-SXM2...  On   | 00000000:89:00.0 Off |                    0 |
| N/A   38C    P0    60W / 300W |  21171MiB / 32480MiB |      0%      Default |
+-------------------------------+----------------------+----------------------+
```

**Comment by williamBlazing** [comment]
@bsuryadevara A couple things to mention. First off when creating `BlazingContext()` by default we are now using the pool allocator, which by default will allocate half the GPU. The memory consumption you are seeing is just that. It has nothing to do with the table. To turn off the pool allocation you can create the context like so:
`bc = BlazingContext(pool=False)`

Additionally, the table you are creating here is based off of a file, which means that the context, is only holding in memory information about the table schema and where the files are, its not actually holding any data. Do `drop_table` in this context would not be apparent as a change in memory consumption. On the other hand if you create a GPU DataFrame (a gdf) in any way, and create a table from it  `bc.create_table('gdfTableName', gdf)` then here the context will maintain a reference to the gdf, so that if the user deletes the gdf, the BlazingSQL context will still have it until you call `drop_table`

**Comment by bsuryadevara** [comment]
@williamBlazing updated as per the suggestion, but it's still  holding `690 MB` memory for `1.4MB` dataset. 
```
import time
import cudf
from blazingsql import BlazingContext

config = {"tables":[{"table_name":"movies", "filepath":"/rapids/my_data/ml/movies.csv"}]}

def run_query(query): 
    bc = BlazingContext(pool=False)
    for table_config in config['tables']:
        table_name = table_config['table_name']
        file_path = table_config['filepath']
        gdf = cudf.read_csv(file_path)
        bc.create_table(table_name, gdf)
    #Run query and get the results
    df = bc.sql(query)
    bc.drop_table("movies")


if __name__ == "__main__":
    for i in range(0, 10):
        run_query('SELECT * FROM main.movies LIMIT 100')
        print("iteration %s completed!"%(i))
        print("it's time to wait for 30secs")
        time.sleep(30)
```

**Comment by felipeblazing** [comment]
When you launch the python process all of the gpu kernels that the process might run also get loaded. The python process itself will consume memory just from that. Now this should not be 600MB but it is about 300 or so.  Can you show us how you are assessing consumption?

**Comment by gumdropsteve** [comment]
@bsuryadevara thanks for filing your 1st BSQL issue & working with the team as it gets sorted out!

Went & tested and noticed this issue when running multiple BlazingContext instances (or deleting or overwriting the same instance) on a single 16GB T4 GPU with (`pool=True`) & without (`pool=False`) pool in Python scripts/Terminal and Jupyter. 

Whichever is started first is where the memory stays. I'm checking memory with `!nvidia-smi`.

#### Replacing instance where pool=True with instance where pool=False
![image](https://user-images.githubusercontent.com/43570913/73433302-5fd05480-42f9-11ea-956a-b38c29e70b2e.png)

#### Deleting instance all together 
![image](https://user-images.githubusercontent.com/43570913/73433377-7d052300-42f9-11ea-902d-7e0a351474a4.png)

#### Starting 2 instances where 1st has pool=False and 2nd has pool=True
![image](https://user-images.githubusercontent.com/43570913/73433428-97d79780-42f9-11ea-93bb-7350d5f2f38a.png)

When script ends or Notebook is restarted the memory is immediately cleared up.

**Comment by bsuryadevara** [comment]
@gumdropsteve thank you for considering this issue as priority. Your explanation make sense. But when I tried in my container it is utilizing `572MB` just for the initialization of the context. do you think this is expected with multi gpu?
And also when I load `1.4MB` dataset it is filling up `690MB` and it is not garbage collected even though function call is done until the demon process exit. 

Please let me know if you have any questions.

![Screen Shot 2020-01-30 at 11 12 25 AM](https://user-images.githubusercontent.com/4812413/73473527-8e2e4e00-4352-11ea-961a-890b15823a30.png)

**Comment by felipeblazing** [comment]
It seems like we might have found a memory leak in CUDF we are diagnosing.

**Comment by gumdropsteve** [comment]
#### Issue confined to Python kernel 
Just found that if I run BlazingContext instances from multiple Jupyter Notebooks (i.e. 1 instance per Notebook) at the same time the initial memory allocation of each instance occurs as expected. Still have the dropping and overwriting/deleting issues, but multiple initial allocations is done successfully.

**Comment by felipeblazing** [comment]
@bsuryadevara So after spending some time looking into . What seems to happen is if a BlazingTable is output from a jupyter cell, as occurs when someone writes
```
bc.create_table(...)
```
Then the reference count is getting incremented and not decremented after this execution causing the table to stay in memory.  I just pushed a fix that prevents it from output from create_table and this resolves the issue you are seeing. I found another potential memory leak (or python GC issue) but the amount of the leak is MUCH smaller. Next nightly release this should be fixed.
